packages/resilience/src/circuit-breaker.ts

```diff
@@ -0,0 +1,114 @@
+import CircuitBreaker from 'opossum';
+// eslint-disable-next-line import/named
+import { AxiosError } from 'axios';
+import { Context, Middleware, MiddlewareIn } from './middleware';
+
+type ErrorFilter = (err) => boolean;
+type KeyBuilder<ContextT extends Context> = (context: ContextT) => string;
+
+function httpErrorFilter(error: AxiosError): boolean {
+  if (
+    error.response?.status &&
+    error.response.status.toString().startsWith('4')
+  ) {
+    return true;
+  }
+  return false;
+}
+
+function circuitBreakerKeyBuilder<ContextT extends Context>(
+  context: ContextT
+): string {
+  return `${context.uri}::${context.tenantId}`;
+}
+
+/**
+ * @internal
+ */
+export function circuitBreakerHttp<
+  ReturnT,
+  ContextT extends Context
+>(): Middleware<ReturnT, ContextT> {
+  return circuitBreaker<ReturnT, ContextT>(
+    circuitBreakerKeyBuilder,
+    httpErrorFilter
+  );
+}
+
+function circuitBreaker<ReturnT, ContextT extends Context>(
+  keyBuilder: KeyBuilder<ContextT>,
+  errorFilter: ErrorFilter
+): Middleware<ReturnT, ContextT> {
+  return (options: MiddlewareIn<any, any>) => () =>
+    (
+      getCircuitBreaker(
+        keyBuilder(options.context),
+        errorFilter
+      ) as CircuitBreaker<any, ReturnT>
+    ).fire(options.fn);
+}
+
+/**
+ * Map of all existing circuit breakers.
+ * Entries are added in a lazy way.
+ * @internal
+ */
+export const circuitBreakers: Record<string, CircuitBreaker> = {};
```

Comment:
[pp] Could you move this variable to the top of the file. I like to have such maps directly visible when you open a file.